interp/builtin.go

```diff
@@ -553,3 +595,45 @@ func (r *Runner) relPath(path string) string {
 	}
 	return filepath.Clean(path)
 }
+
+type getopts struct {
+	argidx  int
+	runeidx int
+}
+
+func (g *getopts) Next(optstr string, args []string) (opt rune, optarg string, done bool) {
+	if len(args) == 0 || g.argidx >= len(args) {
+		return '?', "", true
+	}
+	arg := []rune(args[g.argidx])
+	if len(arg) < 2 || arg[0] != '-' || arg[1] == '-' {
+		return '?', "", true
+	}
+
+	opts := arg[1:]
+	opt = opts[g.runeidx]
+	if g.runeidx+1 < len(opts) {
+		g.runeidx += 1
+	} else {
+		g.argidx += 1
+		g.runeidx = 0
+	}
+
+	i := strings.IndexRune(optstr, opt)
+	if i == -1 {
+		// invalid option
+		return '?', string(opt), false
+	}
+
+	if i+1 < len(optstr) && optstr[i+1] == ':' {
+		if len(args) < g.argidx+1 {
```

Comment:
I'd prefer `if g.argidx >= len(args)`, as that is the common "if I can't use x as an index of y".

Reply:
done